### DISCUSSION TITLE
Long-Term Strategy

### DISCUSSION THREAD

**Discussion by nordlow** [MrSmith33/vox#3] [General]
Are there any plans for scaling and delegating the development of Vox to more developers?

And is there any ordering of priority for implementing the features in `todo.txt`?

**Comment by MrSmith33** [comment]
Scaling is definitely a goal. With linux builds working and tests passing it is now easier for other people to get into development.
I need to spend more time on docs about Vox and implementation.

`todo.txt` is mix of ideas and actual tasks mostly grouped by topic, not priority.

You can consider yourself as a first major contributor 😃 

Edit: I didn't do a major announce on D forums yet.

**Comment by nordlow** [comment]
I could help out improving diagnostics. 😊

**Comment by nordlow** [comment]
How do I build for Linux?

**Comment by nordlow** [comment]
> I could help out improving diagnostics. 😊

And add a flag for syntax and semantic checking only.

**Comment by MrSmith33** [comment]
It is in the readme https://github.com/MrSmith33/tiny_jit#running--testing
For that you would need to add flag to CompilationContext in `context.d` and in `cli.d`. Also pass structure must be changed for that, as pass list is hardcoded rn in `passes.d`.

**Comment by MrSmith33** [comment]
Easy way would be to modify https://github.com/MrSmith33/tiny_jit/blob/master/source/cli.d#L135 to strip unneeded passes.

**Comment by MrSmith33** [comment]
I've split them https://github.com/MrSmith33/tiny_jit/blob/master/source/passes.d#L62-L91, so you need to choose between `commonPasses` and `frontendPasses`.

**Comment by nordlow** [comment]
Ok, thanks. What do you want the flag be called?

BTW: Why do you call this repo `tiny_jit` instead of `vox`? Isn't this Vox's compiler an AOT-compiler?

**Comment by MrSmith33** [comment]
No idea. Probably the same as other compilers.
`Vox` as a name of the lang came much later.

**Comment by MrSmith33** [comment]
I guess the earlier I rename it, the better

**Comment by MrSmith33** [comment]
Let's try `--check-only`

**Comment by nordlow** [comment]
I don't get it. Is there support for Linux ELF output for not? I'm sitting on Linux. And the produced binary on Linux is a Window binary. Linux supports executing Windows binaries now, thought. But still, is there Linux support?

**Comment by nordlow** [comment]
> No idea. Probably the same as other compilers.
> `Vox` as a name of the lang came much later.

Is this a JIT (just-in-time) or AOT (ahead-of-time) compiler? I presume AOT. If so, why the name tiny_jit? Did you plan to design/implement a jitted-language initially?

**Comment by MrSmith33** [comment]
Currently there is no support for building ELF files, only PE files. But on linux you can run in JIT-mode (as tests do)
Initially I toyed with amd64 instruction generation at runtime, hence the name of the repo.

**Comment by nordlow** [comment]
How do I run a hello world in JIT-mode?

**Comment by MrSmith33** [comment]
Use `devtest` version in `main.d`. Select test to run in the `tester.d` in `runDevTests` function

**Comment by MrSmith33** [comment]
@nordlow Are you still interested in implementing `--check-only` yourself?

**Comment by nordlow** [comment]
A bit short on time now. Working 100 percent on new diagnostics for dmd. So go ahead and do it yourself if you feel like. 😉

**Comment by nordlow** [comment]
I could put together an Emacs Flycheck wrapper for Vox when youre done with check only.

What editor are you using?

**Comment by MrSmith33** [comment]
Sublime Text 3
Now that I added --check-only I remember that I have missing return checking in IR generation pass.

**Comment by nordlow** [comment]
> Now that I added --check-only I remember that I have missing return checking

What do you mean by "missing return checking"?

> in IR generation pass.

And why is it in this pass?

**Comment by MrSmith33** [comment]
Check if non-void function is missing a return statement.
It is this check: <https://github.com/MrSmith33/vox/blob/master/source/fe/ast/decl/func.d#L230-L238>
I do it there because I have information if any paths reach the end of the function.

**Comment by nordlow** [comment]
I see. Thanks. How many AST-passes do you do and where are those in the code?

**Comment by nordlow** [comment]
BTW: does Vox require forward references?

**Comment by MrSmith33** [comment]
It is 3 passes + ir gen. You can use `--print-time` for all passes and `--print-mem` for memory

**Comment by MrSmith33** [comment]
no forward references required

**Comment by nordlow** [comment]
> It is 3 passes + ir gen.

How does Vox's passes differ from DMDs?

Is there something Vox can't express but D can and vice versa?

**Comment by MrSmith33** [comment]
No idea, I don't really know dmd's codebase

**Comment by nordlow** [comment]
> No idea, I don't really know dmd's codebase

Ok, thanks.

Is there something Vox will never do but D can because of time/space complexity?

**Comment by nordlow** [comment]
Are there any special kinds of diagnostics you are planning?

**Comment by MrSmith33** [comment]
> Is there something Vox will never do but D can because of time/space complexity?

I didn't find such things yet.
> Are there any special kinds of diagnostics you are planning?

Some static analysis on SSA IR would be nice I think.

**Comment by nordlow** [comment]
> I didn't find such things yet.

Great.

> > Are there any special kinds of diagnostics you are planning?
> Some static analysis on SSA IR would be nice I think.

Do you mean unused variables, duplicate variable writes etc?

What about null pointer derefs?

**Comment by MrSmith33** [comment]
I will add ability to enable do runtime checks at first. Then I may try some global analysis / asserts / annotations / special pointer types etc.

**Comment by MrSmith33** [comment]
But that requires suitable IR

**Comment by nordlow** [comment]
How are run-time checks for pointer deref better than segfault with stack trace a la D?

**Comment by MrSmith33** [comment]
for me D doesn't show stack traces, you need to add segfault handler for that (at least on Windows). With checks you can catch null earlier than with segfault. Just having reliable segfault handler would help.

**Comment by nordlow** [comment]
> for me D doesn't show stack traces, you need to add segfault handler for that (at least on Windows). With checks you can catch null earlier than with segfault.

Can you give an example of such an earlier null catch?

**Comment by MrSmith33** [comment]
When you call method with null this

**Comment by MrSmith33** [comment]
Just made syscalls into usable state. Still a couple of quirks with register allocation to fix, but:
```D
enum u32 stdin = 0;
enum u32 stdout = 1;
enum u32 stderr = 2;

@extern(syscall, 60)
void exit(i32 error_code);

@extern(syscall, 1)
void sys_write(u32 fd, u8* buf, u64 count);

void write(u32 fd, u8[] data) {
    sys_write(fd, data.ptr, data.length);
}

void main() {
    write(stdout, "hello linux\n");
    exit(0);
}
```
```
$ ./tjc ./hello_linux.vx
$ ./hello_linux.exe
hello linux\n
```
Need to implement escape sequences in strings.

**Comment by nordlow** [comment]
What are your plans for D-style ranges or Rust-style iterators? I'd like to see Vox picking a Rust-like iterator API [1] with, say, `next` returning an optional `T?`.

[1] https://doc.rust-lang.org/std/iter/trait.Iterator.html

**Comment by MrSmith33** [comment]
For now I like Jai-style iterators, which are done via macros. But they can't passed as a value like ranges can, so composability is worse.
Do Rust iterators support random access, bidirectionality etc features of D ranges?

**Comment by nordlow** [comment]
> For now I like Jai-style iterators, which are done via macros.

Do you have link to Jai-style iterators? Why macros?

> But they can't passed as a value like ranges can, so composability is worse.

That's in, imho, a show-stopper.

> Do Rust iterators support random access, bidirectionality etc features of D ranges?
random-access: https://doc.rust-lang.org/1.0.0/std/iter/trait.RandomAccessIterator.html
bidirectional: https://doc.rust-lang.org/1.0.0/std/iter/trait.DoubleEndedIterator.html

Note that a Rust-iterator is similar to a range in D and C++20.

**Comment by MrSmith33** [comment]
Is there any in-depth comparison between D-style ranges and Rust ones?

**Comment by nordlow** [comment]
I haven't seen any. Rust has no concept of infinite ranges. There might be more differences.
